src/pudl_archiver/depositors/zenodo/depositor.py

```diff
@@ -0,0 +1,609 @@
+"""Handle all deposition actions within Zenodo."""
+
+import asyncio
+import importlib
+import json
+import logging
+import os
+import traceback
+from hashlib import md5
+from pathlib import Path
+from typing import BinaryIO, Literal
+
+import aiohttp
+import semantic_version  # type: ignore  # noqa: PGH003
+import yaml
+from pydantic import BaseModel, PrivateAttr
+
+from pudl_archiver.depositors.depositor import (
+    DepositionAction,
+    DepositionChange,
+    DraftDeposition,
+    PublishedDeposition,
+)
+from pudl_archiver.frictionless import DataPackage, ResourceInfo
+from pudl_archiver.utils import RunSettings, Url, retry_async
+
+from .entities import (
+    Deposition,
+    DepositionMetadata,
+    Doi,
+    Record,
+    SandboxDoi,
+)
+
+logger = logging.getLogger(f"catalystcoop.{__name__}")
+
+
+def _compute_md5(file_path: Path) -> str:
+    """Compute an md5 checksum to compare to files in zenodo deposition."""
+    hash_md5 = md5()  # noqa: S324
+    with Path.open(file_path, "rb") as f:
+        for chunk in iter(lambda: f.read(4096), b""):
+            hash_md5.update(chunk)
+
+    return hash_md5.hexdigest()
+
+
+class ZenodoClientError(Exception):
+    """Captures the JSON error information from Zenodo."""
+
+    def __init__(self, status, message, errors=None):
+        """Constructor.
+
+        Args:
+            status: status message of response
+            message: message of response
+            errors: if any, list of errors returned by response
+        """
+        self.status = status
+        self.message = message
+        self.errors = errors
+
+    def __str__(self):
+        """Cast to string."""
+        return repr(self)
+
+    def __repr__(self):
+        """But the kwargs are useful for recreating this object."""
+        return f"ZenodoClientError(status={self.status}, message={self.message}, errors={self.errors})"
+
+
+class DatasetSettings(BaseModel):
+    """Simple model to validate doi's in settings."""
+
+    production_doi: Doi | None = None
+    sandbox_doi: SandboxDoi | None = None
+
+
+class ZenodoBaseDepositionInterface(BaseModel):
```

Comment:
Yeah I used inheritance here because I wanted to avoid some boiler plate, but ultimately this approach felt kind of tricky/error prone, so I'm good with using composition instead.